### ISSUE TITLE
A test class with Java 9+ dependencies can not be disabled with @DisabledOnJre(JRE.JAVA_8)

### ISSUE THREAD

**Issue by hgschmie** [junit-team/junit5#2751]
I have a project that is compiled to Java 8 compatibility. However, there are some test dependencies that no longer support Java 8 but only newer versions (Java 9+). 

```java
public class TestWithAJava11Dependency {
    @RegisterExtension
    public static Java11OnlyExtension = Java11OnlyExtension.instance();

    ....
}
```

My naive idea was to decorate this with `@DisabledOnJre(JRE.JAVA_8)`. However, when running this code on a Java 8 JRE, it reports

```
java.lang.UnsupportedClassVersionError: .../Java11OnlyExtension has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up to 52.0
```

so even though the code itself is compiled to java 8 and it is annotated with "skip when on java 8", the class is loaded and fails on Java 8

## Deliverables

- [ ] Clearly it would be good if the test class could somehow be inspected without the Java 9+ dependencies being loaded by the class loader. I am not sure if that is possible; in any case it would be reasonable to document this clearly.

**Comment by hgschmie** [comment]
completely understand. What I was trying to illustrate was that this annotation can not be used to e.g. disable tests based on runtime when the runtime can not load the test class and that is a problem for cross-compilation and testing which is IMHO a common use case.

One possible solution that I can see would be e.g. allowing to add this to the package-info.java file as an annotation on the package statement and then not even attempting to load the test class:

```java
@DisabledTestsOnJre(JRE.JAVA_8, { TestClassA.class, TestClassB.class })
```

**Comment by hgschmie** [comment]
So I spent a little bit of time with the code and here is what I see:

- beforeEach (in my extension) is called 


- When the JupiterTestEngine starts, it creates a new engine descriptor in the `discover(EngineDiscoveryRequest, UniqueId)` method. This has all the test methods correctly discovered, even on JDK8. So at this stage, the engine can completely inspect all the test classes. 
- The created `ClassTestDescriptor` even has the right annotation (source -> javaClass -> annotationData -> declaredAnnotations`
- However, the code ultimately recurses through the test containers until it ends up with a NodeTestTask which contains the following code in the `execute()` method:

```java
prepare();
if (throwableCollector.isEmpty()) {
    checkWhetherSkipped();
}
```

and that is unfortunate. The error happens when calling `registerExtensionFromFields` which is called from `prepare()`. IAW, the code loads the test class first and then checks whether it needs to skip the test. 

I am not sure why that is. It seems to me that there may be cases where a condition is defined in the test class, but that seems rare. It should be possible to manage at least the most common conditions not by reflecting on the test classes themselves but by evaluating the information present in the ClassTestDescriptor.

**Comment by marcphilipp** [comment]
In order to check whether a `TestDescriptor` should be skipped all extensions need to be registered (since `ExecutionCondition` extensions determine whether it's skipped).

Could you use `@Tag` to mark those tests and skip them?

**Comment by hgschmie** [comment]
Hi @marcphilipp. Thank you for getting back quickly. 

Yes, I agree that this is a hen-egg situation. However, as this will pretty much only affect the "based on the JRE version" checks (because those may be affected by the inability to load the class into the JVM), would it be possible to "prerun" these checks before we actually load the class into memory?

**Comment by hgschmie** [comment]
Tagging only works to a point, BTW. 

It falls down when it comes to nested test classes as the code will at some point during discovery (!) (so long before tags are evaluated) try to determine if a class has nested tests and in that case not do the careful inspection that it does to discover tests but simply call `ReflectionUtils.findNestedClasses()` which in turn tries to load the class (which then immediately explodes with `UnsupportedClassVersionError`). I guess that would probably is a bug. (I tried using a Tag both on the enclosing class and each of the nested classes. Neither works.

**Comment by marcphilipp** [comment]
Thanks for trying out tags. I had not thought about class loading still taking place during discovery. I don't think that's a bug but having classes that can't be loaded on the test JVM is rather an edge case.

May I ask why are you attempting to run these tests on Java 8 at all if they don't actually support that release?

The other idea I had was creating a lazy wrapper of the problematic extension that skips tests that use it on Java 8.

**Comment by hgschmie** [comment]
We are using Junit 5 for unit tests in the jdbi project (https://jdbi.org/). Our code is compiled to Java 8 byte code but at this point, we do require Java 11+ for the tool chain. 

With the deprecation of Java 8, we moved the status of Java 8 support to "mature, supported, best effort" and placed our base line for testing at Java 11. As a result, we started to modernize some of the libraries that we use in the unit tests (that are not part of the code itself but dependencies). Some of those (such as caffeine) have moved to a Java 11 base line as well. So we can run all the tests find on a JDK 11, but for a JDK8, we need to skip those. right now, we use the maven surefire plugin exclusions but that is cumbersome and error prone. Hence the idea of either tagging or annotating the test classes that no longer run under Java 8 (because they use Java 11 dependencies) so that they get skipped.
